src/Plugin.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE GADTs #-}
+
+module Plugin (plugin) where
+
+import Control.Concurrent.STM
+import Control.Exception
+import Control.Monad
+import Data.IORef
+import GHC
+import GHC.Driver.Hooks
+import GHC.Driver.Pipeline
+import GHC.Driver.Pipeline.Phases
+import GHC.Plugins as Plugins
+import GHC.Types.Name.Cache
+import HieDb.Create
+import HieDb.Types
+import System.Directory (doesPathExist)
+import System.FilePath
+import qualified System.IO.Unsafe as Unsafe
+
+plugin :: Plugin
+plugin =
+  defaultPlugin
+    { pluginRecompile = Plugins.purePlugin
+    , driverPlugin = driver
+    }
+
+driver :: [CommandLineOption] -> HscEnv -> IO HscEnv
+driver _ hscEnv = do
+  initializeHiedb
+  pure
+    hscEnv
+      { hsc_hooks =
+          (hsc_hooks hscEnv)
+            { runPhaseHook = Just phaseHook
+            }
+      }
+ where
+  initializeHiedb = liftIO $ withHieDb defaultHiedbFile initConn
+
+  -- We index using a phase hook instead of typeCheckResultAction since
+  -- the hie file can be written after that plugin phase
+  phaseHook =
+    PhaseHook $ \phase -> do
+      case phase of
+        T_HscPostTc _ modSummary _ _ _ -> do
+          let dynFlags = hsc_dflags hscEnv
+              hieDirectory = hieDir dynFlags
+          _ <- liftIO $ addModuleToDb defaultHiedbFile (ms_mod modSummary) hieDirectory
```

Comment:
It's possible to use `withAsync` or `forkIO` to do this asynchronously here but I've found at a high number of modules that actually incurs a performance cost. Using a scheduling thread and queueing might work but can run into issues since the parent can get killed before the child thread is done indexing - perhaps using `forkProcess` with a queue could work too but I haven't tried it. Empirically though this is pretty fast.